src/include/platform/internal/GenericPlatformManagerImpl_POSIX.ipp

```diff
@@ -128,15 +128,35 @@ CHIP_ERROR GenericPlatformManagerImpl_POSIX<ImplClass>::_StartChipTimer(System::
     return CHIP_NO_ERROR;
 }
 
+#if CHIP_SYSTEM_CONFIG_USE_LIBEV
+template <class ImplClass>
+void GenericPlatformManagerImpl_POSIX<ImplClass>::_DispatchEventViaScheduleWork(System::Layer * aLayer, void * appState)
+{
+    const ChipDeviceEvent * event = static_cast<const ChipDeviceEvent *>(appState);
+    PlatformMgrImpl().DispatchEvent(event);
+    delete event;
+}
+#endif // CHIP_SYSTEM_CONFIG_USE_LIBEV
+
 template <class ImplClass>
 CHIP_ERROR GenericPlatformManagerImpl_POSIX<ImplClass>::_PostEvent(const ChipDeviceEvent * event)
 {
+#if CHIP_SYSTEM_CONFIG_USE_LIBEV
+    // Schedule dispatching this event via System Layer's ScheduleWork
+    ChipDeviceEvent * eventCopyP = new ChipDeviceEvent;
+    VerifyOrDie(eventCopyP != nullptr);
+    *eventCopyP = *event;
+    SystemLayer().ScheduleWork(&_DispatchEventViaScheduleWork, eventCopyP);
```

Comment:
SystemLayer::ScheduleWork is documented as only being safe to call with the Matter lock held, while PostEvent is documented to be callable from any thread without holding any locks.  How can this method delegate to that one, exactly?

Reply:
You're right of course, when there are threads on the platform level that may use `PostEvent`, this does not work.

However, the reason for implanting libev support into matter in the first place is exactly to get rid of threads, and have a single-threaded app based on libev including all of matter, for those who need that (my case).

So I wonder how to proceed: try to add a lock (similar to DeviceSafeQueue) or rather an assertion that would prevent issuing events from non-mainthreads completely?

Reply:
There's already an assertion inside `SystemLayer().ScheduleWork()`.  So probably just documenting here why this is OK would be fine.

Reply:
> There's already an assertion inside `SystemLayer().ScheduleWork()`

I don't see an assertion that would prevent calling from the wrong thread, just those that ensure there is a mainloop; am I looking in the wrong place? (`SystemLayerImplSelect.cpp`)

`PrepareEvents()` e.g. calls `assertChipStackLockedByCurrentThread()`, this could be called in `ScheduleWork()` in the libev case, maybe?

Reply:
`LayerImplSelect::ScheduleWork` calls `assertChipStackLockedByCurrentThread();` right at the top, no?

Reply:
It would go in the libev case in the impl of PostEvent.  Which should perhaps do its own assert to make things really clear.